elastic/kibana
# [filters] filter operator registry and spatial_filter operator

Fixes https://github.com/elastic/kibana/issues/109340 and https://github.com/elastic/kibana/issues/107044

### Background
In Maps, users can [create spatial filters](https://www.elastic.co/guide/en/kibana/current/maps-create-filter-from-map.html). The filter is added to the Filter bar.
<img width="300" alt="Screen Shot 2021-09-03 at 4 07 33 PM" src="https://user-images.githubusercontent.com/373691/133142628-9b03dec2-69ee-4c7c-ac8d-841168c88a94.png">
There are 2 important properties about spatial filters created from a Map:
* `filter.meta.type` is set to `spatial_filter`. Maps uses `filter.meta.type` to identify spatial filters. Spatial filters are displayed on the map to provide a visual reference of all applied spatial filters.
* `filter.meta.isMultiIndex` is set to true since Elasticsearch Query DSL for the filter is not tied to a single index pattern or field. The Elasticsearch Query DSL is generated in such a way that the filter works for all map layers, regardless of index pattern or spatial field. In the screenshot above a single filter is filtering both layers, even though they are from different index patterns and different spatial fields.
  * Each layer identifies one or more geo_point or geo_shape fields that is used to display features for that layer
  * The Elasticsearch Query DSL combines these into a `bool.should` clause where each field provides a `bool.must` clause ensuring the spatial field exists and that the document satisfies the spatial filter. Any document matching any of the `bool.must` clauses allows the bool.should clause to produce a match.

### Test instructions
* Install web logs sample data set and open "[Logs] Total Requests and Bytes" map
* Click wrench and select "Draw bounds to filter data". Draw bounds and create a filter.
* In the Filter bar, edit the filter. Notice that the filter editor opens in "values" view and not the "query DSL" view. Notice that Index pattern select, Field select, and operator select are not displayed.
    <img width="300" alt="Screen Shot 2021-09-13 at 1 30 41 PM" src="https://user-images.githubusercontent.com/373691/133144989-9a6f6cc6-37e3-407a-bfb0-c97464943005.png">
* Change the label and click save. The filter should still be displayed on the map.

### Changes to Data plugin
* Updates FilterEditor to handle multi index filters. When a filter is a multi index filter, the editor does not display IndexPattern select, Field select, or operator select.
* Creates a filter operator registry so that the maps plugin can register a filter operator for `spatial_filter` type.
* Migrates hard coded filter operator array to filter operator registry. This involved refactoring several places that used a switch on `meta.type` and moving the logic into Operator interface.
* Updates setup contract to expose `filterEditor` service.

### Changes to Maps plugin
  * Registers a spatialFilterOperator. Now, when editing a spatial filter, the filter editor shows "values" form instead of the "query DSL" form. This allows for changing the filter label. In the future, there will be UI showing the filter on a map and allowing users to edit the filter shape or redraw it.
  * Updates spatial filter creation to set `filter.meta.isMultiIndex = true` for all generated filters even if there is only a single geo field. The reason for this is that geo field names are decoupled from index patterns and there is no way of knowing if a geo field name exists on multi index patterns. In practice, any spatial filter created from the map is not coupled to field/index, thus its a multi index filter.
  * Creates a folder `setup` to organize files used during setup phase since `public` folder was getting a bit unorganized.
